# Case 1: 点球成金
## 加载和浏览数据集

In [1]:
baseball = read.csv("../../datas/Moneyball.csv")
str(baseball)

'data.frame':	932 obs. of  15 variables:
 $ Team        : chr  "ANA" "ARI" "ATL" "BAL" ...
 $ League      : chr  "AL" "NL" "NL" "AL" ...
 $ Year        : int  2002 2002 2002 2002 2002 2002 2002 2002 2002 2002 ...
 $ RS          : int  851 819 708 667 859 706 856 709 739 778 ...
 $ RA          : int  644 674 565 773 665 759 798 774 837 898 ...
 $ W           : int  99 98 101 67 93 67 81 78 74 73 ...
 $ OBP         : num  0.341 0.346 0.331 0.309 0.345 0.321 0.338 0.33 0.321 0.337 ...
 $ SLG         : num  0.433 0.423 0.409 0.403 0.444 0.413 0.449 0.408 0.412 0.423 ...
 $ BA          : num  0.282 0.267 0.26 0.246 0.277 0.246 0.268 0.253 0.249 0.274 ...
 $ Playoffs    : int  1 1 1 0 0 0 0 0 0 0 ...
 $ RankSeason  : int  3 4 2 NA NA NA NA NA NA NA ...
 $ RankPlayoffs: int  1 4 4 NA NA NA NA NA NA NA ...
 $ G           : int  162 162 161 162 162 162 162 162 162 162 ...
 $ OOBP        : num  0.314 0.305 0.313 0.336 0.308 0.331 0.33 0.338 0.348 0.349 ...
 $ OSLG        : num  0.392 0.397 0.364

由于我们确认了Moneyball的索赔，因此我们希望使用Paul DePodesta在2002年获得的数据来构建模型。

In [2]:
moneyball = subset(baseball, Year < 2002)
str(moneyball)

'data.frame':	902 obs. of  15 variables:
 $ Team        : chr  "ANA" "ARI" "ATL" "BAL" ...
 $ League      : chr  "AL" "NL" "NL" "AL" ...
 $ Year        : int  2001 2001 2001 2001 2001 2001 2001 2001 2001 2001 ...
 $ RS          : int  691 818 729 687 772 777 798 735 897 923 ...
 $ RA          : int  730 677 643 829 745 701 795 850 821 906 ...
 $ W           : int  75 92 88 63 82 88 83 66 91 73 ...
 $ OBP         : num  0.327 0.341 0.324 0.319 0.334 0.336 0.334 0.324 0.35 0.354 ...
 $ SLG         : num  0.405 0.442 0.412 0.38 0.439 0.43 0.451 0.419 0.458 0.483 ...
 $ BA          : num  0.261 0.267 0.26 0.248 0.266 0.261 0.268 0.262 0.278 0.292 ...
 $ Playoffs    : int  0 1 1 0 0 0 0 0 1 0 ...
 $ RankSeason  : int  NA 5 7 NA NA NA NA NA 6 NA ...
 $ RankPlayoffs: int  NA 1 3 NA NA NA NA NA 4 NA ...
 $ G           : int  162 162 162 162 161 162 162 162 162 162 ...
 $ OOBP        : num  0.331 0.311 0.314 0.337 0.329 0.321 0.334 0.341 0.341 0.35 ...
 $ OSLG        : num  0.412 0.404 0.384 0.

## 预测胜利
我们首先建立一个线性回归模型，以使用得分数和失分数之间的差异来预测获胜。为了使操作更容易一些，让我们从创建一个名为RD的新变量RD开始，并将其设置为的分数减去失分数

In [3]:
moneyball$RD = moneyball$RS - moneyball$RA
str(moneyball)

'data.frame':	902 obs. of  16 variables:
 $ Team        : chr  "ANA" "ARI" "ATL" "BAL" ...
 $ League      : chr  "AL" "NL" "NL" "AL" ...
 $ Year        : int  2001 2001 2001 2001 2001 2001 2001 2001 2001 2001 ...
 $ RS          : int  691 818 729 687 772 777 798 735 897 923 ...
 $ RA          : int  730 677 643 829 745 701 795 850 821 906 ...
 $ W           : int  75 92 88 63 82 88 83 66 91 73 ...
 $ OBP         : num  0.327 0.341 0.324 0.319 0.334 0.336 0.334 0.324 0.35 0.354 ...
 $ SLG         : num  0.405 0.442 0.412 0.38 0.439 0.43 0.451 0.419 0.458 0.483 ...
 $ BA          : num  0.261 0.267 0.26 0.248 0.266 0.261 0.268 0.262 0.278 0.292 ...
 $ Playoffs    : int  0 1 1 0 0 0 0 0 1 0 ...
 $ RankSeason  : int  NA 5 7 NA NA NA NA NA 6 NA ...
 $ RankPlayoffs: int  NA 1 3 NA NA NA NA NA 4 NA ...
 $ G           : int  162 162 162 162 161 162 162 162 162 162 ...
 $ OOBP        : num  0.331 0.311 0.314 0.337 0.329 0.321 0.334 0.341 0.341 0.35 ...
 $ OSLG        : num  0.412 0.404 0.384 0.

In [4]:
PredictWins = lm(W ~ RD, data = moneyball)
summary(PredictWins)


Call:
lm(formula = W ~ RD, data = moneyball)

Residuals:
     Min       1Q   Median       3Q      Max 
-14.2662  -2.6509   0.1234   2.9364  11.6570 

Coefficients:
             Estimate Std. Error t value Pr(>|t|)    
(Intercept) 80.881375   0.131157  616.67   <2e-16 ***
RD           0.105766   0.001297   81.55   <2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 3.939 on 900 degrees of freedom
Multiple R-squared:  0.8808,	Adjusted R-squared:  0.8807 
F-statistic:  6651 on 1 and 900 DF,  p-value: < 2.2e-16


## 预测得分 
我们想看看是否可以使用线性回归来使用三个统计数据(上垒率（OBP），长打率（SLG）和打击率（BA）)来预测得分（RS）。

In [5]:
PredictRS = lm(RS ~ OBP + SLG + BA, data = moneyball)
summary(PredictRS)


Call:
lm(formula = RS ~ OBP + SLG + BA, data = moneyball)

Residuals:
    Min      1Q  Median      3Q     Max 
-70.941 -17.247  -0.621  16.754  90.998 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)  -788.46      19.70 -40.029  < 2e-16 ***
OBP          2917.42     110.47  26.410  < 2e-16 ***
SLG          1637.93      45.99  35.612  < 2e-16 ***
BA           -368.97     130.58  -2.826  0.00482 ** 
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 24.69 on 898 degrees of freedom
Multiple R-squared:  0.9302,	Adjusted R-squared:   0.93 
F-statistic:  3989 on 3 and 898 DF,  p-value: < 2.2e-16


根据定义（您可以在网上找到这些定义），这三个统计数据彼此非常相似，它们可以衡量玩家的击球表现。这些统计数据越大，球员的击球表现就越好。因此，我们希望所有这些变量的系数都为正，因为更好的参与者应该导致更多的得分。BA的违反直觉的负系数表明该模型可能存在问题。如果我们只关心模型预测的准确性，那么模型就很好。但是，在这种情况下，团队依赖模型聘用潜在的参与者，因此错误的系数将导致严重的错误决策。   

考虑到这些统计数据相互之间密切相关，因为它们都以定义稍有不同的方式来衡量击球表现，因此我们怀疑模型中可能存在一些多重共线性问题。我们可以检查这三个变量之间的相关性以确认这一点。

In [6]:
cor(data.frame(moneyball$OBP, moneyball$SLG, moneyball$BA))

moneyball.OBP moneyball.SLG moneyball.BA
moneyball.OBP 1.0000000     0.8061539     0.8540549   
moneyball.SLG 0.8061539     1.0000000     0.8140681   
moneyball.BA  0.8540549     0.8140681     1.0000000

我们可以看到这三个变量确实彼此高度相关，因为所有相关系数都大于0.8。要决定保留哪个变量，我们需要执行更多分析。

首先，让我们分别对这三个变量进行单变量分析，以检查其预测能力。通过单变量分析，即运行单变量线性回归。

In [7]:
PredictRS1 = lm(RS ~ OBP, data = moneyball)
summary(PredictRS1)


Call:
lm(formula = RS ~ OBP, data = moneyball)

Residuals:
     Min       1Q   Median       3Q      Max 
-118.317  -26.729    1.211   26.849  132.115 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept) -1079.02      27.98  -38.56   <2e-16 ***
OBP          5486.30      86.01   63.78   <2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 39.74 on 900 degrees of freedom
Multiple R-squared:  0.8189,	Adjusted R-squared:  0.8187 
F-statistic:  4069 on 1 and 900 DF,  p-value: < 2.2e-16


In [8]:
PredictRS1 = lm(RS ~ OBP, data = moneyball)
summary(PredictRS1)


Call:
lm(formula = RS ~ OBP, data = moneyball)

Residuals:
     Min       1Q   Median       3Q      Max 
-118.317  -26.729    1.211   26.849  132.115 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept) -1079.02      27.98  -38.56   <2e-16 ***
OBP          5486.30      86.01   63.78   <2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 39.74 on 900 degrees of freedom
Multiple R-squared:  0.8189,	Adjusted R-squared:  0.8187 
F-statistic:  4069 on 1 and 900 DF,  p-value: < 2.2e-16


In [9]:
PredictRS2 = lm(RS ~ SLG, data = moneyball)
summary(PredictRS2)


Call:
lm(formula = RS ~ SLG, data = moneyball)

Residuals:
     Min       1Q   Median       3Q      Max 
-103.113  -24.343   -1.906   22.544  126.598 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)  -315.51      13.86  -22.76   <2e-16 ***
SLG          2610.88      35.38   73.80   <2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 35.16 on 900 degrees of freedom
Multiple R-squared:  0.8582,	Adjusted R-squared:  0.858 
F-statistic:  5446 on 1 and 900 DF,  p-value: < 2.2e-16


In [10]:
PredictRS3 = lm(RS ~ BA, data = moneyball)
summary(PredictRS3)


Call:
lm(formula = RS ~ BA, data = moneyball)

Residuals:
     Min       1Q   Median       3Q      Max 
-153.815  -37.476   -2.814   34.911  184.236 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)  -806.35      33.66  -23.96   <2e-16 ***
BA           5849.88     130.21   44.92   <2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 51.85 on 900 degrees of freedom
Multiple R-squared:  0.6916,	Adjusted R-squared:  0.6913 
F-statistic:  2018 on 1 and 900 DF,  p-value: < 2.2e-16


上面的单变量分析表明，单独使用OBP和SLG可以得到更高的$R^2$. OBP的系数远大于SLG。让我们用OBP和SLG进行多元线性回归来确认发现。

In [11]:
PredictRS4 = lm(RS ~ SLG + OBP, data = moneyball)
summary(PredictRS4)


Call:
lm(formula = RS ~ SLG + OBP, data = moneyball)

Residuals:
    Min      1Q  Median      3Q     Max 
-70.838 -17.174  -1.108  16.770  90.036 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)  -804.63      18.92  -42.53   <2e-16 ***
SLG          1584.91      42.16   37.60   <2e-16 ***
OBP          2737.77      90.68   30.19   <2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 24.79 on 899 degrees of freedom
Multiple R-squared:  0.9296,	Adjusted R-squared:  0.9294 
F-statistic:  5934 on 2 and 899 DF,  p-value: < 2.2e-16


我们可以观察到, 这一模型的$R^2$近似于有三个因子变量的$R^2$ :$0.9302$.  换句话说，BA的效果几乎可以被OBP和SLG捕获。如Moneyball所述，OBP和SLG的系数确定了变量的相对重要性。如果进一步的分析,会发现其他组合的$R^2$将大大小于0.9302

我们决定保留`OBP`和`SLG`，即使它们彼此高度相关，这有两个主要原因. 一, 使用两个因子的模型的$R^2$高于使用单个因子的模型的$R^2$,这表明每个变量中都有大量信息无法被另一个变量捕获。其次，两个变量的系数均显着为正，并且其大小与单变量分析一致。

## 预测失分
我们可以进行类似的线性回归分析，以构建模型来预测失分.

In [12]:
PredictRA = lm(RA ~ OOBP + OSLG, data = moneyball)
summary(PredictRA)


Call:
lm(formula = RA ~ OOBP + OSLG, data = moneyball)

Residuals:
    Min      1Q  Median      3Q     Max 
-82.397 -15.178  -0.129  17.679  60.955 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)  -837.38      60.26 -13.897  < 2e-16 ***
OOBP         2913.60     291.97   9.979 4.46e-16 ***
OSLG         1514.29     175.43   8.632 2.55e-13 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 25.67 on 87 degrees of freedom
  (812 observations deleted due to missingness)
Multiple R-squared:  0.9073,	Adjusted R-squared:  0.9052 
F-statistic: 425.8 on 2 and 87 DF,  p-value: < 2.2e-16
